File: arkav8.0/E.py
from collections import defaultdict

n, x = map(int, input().split())
val = list(map(int, input().split()))

tree = defaultdict(list)
for _ in range(n-1):
    u, v = map(int, input().split())
    tree[u].append(v)
    tree[v].append(u)

vis = set()
q = [x]
while len(q) > 0:
    cur = q.pop(0)
    vis.add(cur)
    for child in tree[cur]:
        if child not in vis:
            val[child-1] = min(val[child-1], val[cur-1])
            q.append(child)
print(" ".join(map(str, val)))

File: arkav8.0/E.cpp
#include <iostream>
#include <unordered_map>
#include <set>
#include <vector>
#include <queue>

using namespace std;

int main() {
    int n, x;
    cin >> n >> x;

    vector<int> val(n);
    for (int i = 0; i < n; i++) {
        cin >> val[i];
    }

    unordered_map<int, vector<int>> tree;
    for (int i = 0; i < n - 1; i++) {
        int u, v;
        cin >> u >> v;
        tree[u].push_back(v);
        tree[v].push_back(u);
    }

    set<int> vis;
    queue<int> q;
    q.push(x);
    while (!q.empty()) {
        int cur = q.front();
        q.pop();
        vis.insert(cur);
        for (int child : tree[cur]) {
            if (vis.count(child) == 0) {
                val[child - 1] = min(val[child - 1], val[cur - 1]);
                q.push(child);
            }
        }
    }

    for (int i = 0; i < n; i++) {
        cout << val[i] << " ";
    }
    cout << endl;

    return 0;
}
